File: VU_KHUSHI/DataStructure/2/Python/QueueArray.py
class QueueArray:
    def __init__(self):
        self.arr = [None] * 100
        self.begin = -1
        self.end = -1
        self.size_of_queue = 0

    def enqueue(self, data):
        if self.is_full():
            print(f"The Queue is full. Cannot enqueue {data}!")
            return -1
        if self.is_empty():
            self.begin = 0
        self.end = (self.end + 1) % 100
        self.arr[self.end] = data
        self.size_of_queue += 1
        print(f"{data} is successfully inserted at index {self.end} in the Queue!")
        return self.end

    def is_full(self):
        return self.size_of_queue == 100

    def is_empty(self):
        return self.size_of_queue == 0

    def dequeue(self):
        if self.is_empty():
            print("The Queue is empty. Cannot dequeue!")
            return -1
        dequeued_value = self.arr[self.begin]
        self.begin = (self.begin + 1) % 100
        self.size_of_queue -= 1
        if self.size_of_queue == 0:
            self.begin = -1
            self.end = -1
        print(f"Dequeued: {dequeued_value}")
        return dequeued_value

    def get_front(self):
        if self.is_empty():
            print("There is no element in the Queue!")
            return -1
        print(f"Front element: {self.arr[self.begin]} at index {self.begin}")
        return self.arr[self.begin]

    def get_rear(self):
        if self.is_empty():
            print("There is no element in the Queue!")
            return -1
        print(f"Rear element: {self.arr[self.end]} at index {self.end}")
        return self.arr[self.end]

    def size(self):
        print(f"Number of elements in the Queue: {self.size_of_queue}")
        return self.size_of_queue

    def print_queue(self):
        if self.is_empty():
            print("The Queue is empty!")
            return
        print("[", end=" ")
        for i in range(self.size_of_queue):
            print(self.arr[(self.begin + i) % 100], end=", ")
        print("]")

# Example usage:
queue = QueueArray()
queue.enqueue(10)
queue.enqueue(20)
queue.enqueue(30)
queue.print_queue()
queue.dequeue()
queue.get_front()
queue.get_rear()
queue.size()

File: VU_KHUSHI/DataStructure/2/JavaScript/QueueArray.js
class QueueArray {
    constructor() {
        this.arr = new Array(100);
        this.begin = -1;
        this.end = -1;
        this.sizeOfQueue = 0;
    }

    enqueue(data) {
        if (this.isFull()) {
            console.log(`The Queue is full. Cannot enqueue ${data}!`);
            return -1;
        }
        if (this.isEmpty()) {
            this.begin = 0;
        }
        this.end = (this.end + 1) % 100;
        this.arr[this.end] = data;
        this.sizeOfQueue++;
        console.log(`${data} is successfully inserted at index ${this.end} in the Queue!`);
        return this.end;
    }

    isFull() {
        return this.sizeOfQueue === 100;
    }

    isEmpty() {
        return this.sizeOfQueue === 0;
    }

    dequeue() {
        if (this.isEmpty()) {
            console.log("The Queue is empty. Cannot dequeue!");
            return -1;
        }
        const dequeuedValue = this.arr[this.begin];
        this.begin = (this.begin + 1) % 100;
        this.sizeOfQueue--;
        if (this.sizeOfQueue === 0) {
            this.begin = -1;
            this.end = -1;
        }
        console.log(`Dequeued: ${dequeuedValue}`);
        return dequeuedValue;
    }

    getFront() {
        if (this.isEmpty()) {
            console.log("There is no element in the Queue!");
            return -1;
        }
        console.log(`Front element: ${this.arr[this.begin]} at index ${this.begin}`);
        return this.arr[this.begin];
    }

    getRear() {
        if (this.isEmpty()) {
            console.log("There is no element in the Queue!");
            return -1;
        }
        console.log(`Rear element: ${this.arr[this.end]} at index ${this.end}`);
        return this.arr[this.end];
    }

    size() {
        console.log(`Number of elements in the Queue: ${this.sizeOfQueue}`);
        return this.sizeOfQueue;
    }

    printQueue() {
        if (this.isEmpty()) {
            console.log("The Queue is empty!");
            return;
        }
        let output = "[ ";
        for (let i = 0; i < this.sizeOfQueue; i++) {
            output += `${this.arr[(this.begin + i) % 100]}, `;
        }
        output = output.slice(0, -2) + " ]"; // Remove trailing comma and space
        console.log(output);
    }
}

// Example usage:
const queue = new QueueArray();
queue.enqueue(10);
queue.enqueue(20);
queue.getFront();
queue.getRear();
queue.printQueue();
queue.dequeue();
queue.size();
queue.printQueue();
